mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp

```diff
@@ -1937,6 +2069,8 @@ void mlir::linalg::populateFoldReshapeOpsByExpansionPatterns(
     const ControlFusionFn &controlFoldingReshapes) {
   patterns.add<FoldReshapeWithGenericOpByExpansion>(patterns.getContext(),
                                                     controlFoldingReshapes);
+  patterns.add<FoldPadWithProducerReshapeOpByExpansion>(patterns.getContext(),
```

Comment:
AFAICS both the patterns added here are "propagating by collapsing". You are moving the expand_shape down and collapse_shape up. In both cases the pad is happening on collapsed dimensions. So you should add both to the `populateFoldReshapeOpsByCollapsingPatterns`.

Reply:
`FoldPadWithProducerReshapeOpByExpansion` pushes the producer collapse_shape down, expanding the pad. `FoldPadWithProducerReshapeOpByCollapsing` pushes the producer expand_shape down, collapsing the pad. I think these are in the right places here.

Reply:
Ok, I twisted myself in a knot when looking at tests. You are right.